File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.js
let fs = require('fs');
let input = fs.readFileSync(0, 'utf8').trim().split('\n');

const [K, N] = input[0].split(' ').map(Number);
const lens = input.slice(1).map(Number);

function solution(K, N, lens) {
  let left = 1;
  let right = Math.max(...lens);

  while (left <= right) {
    let mid = ~~((left + right) / 2);
    let count = 0;
    lens.forEach((len) => {
      count += ~~(len / mid);
    });

    if (count >= N) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return right;
}

console.log(solution(K, N, lens));

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
[K, N] = map(int, input().split())
lines = []
for _ in range(K):
    lines.append(int(input()))

left = 1
right = max(lines)

while left <= right:
    mid = (left + right) // 2
    count = 0
    for line in lines:
        count += line // mid

    if count < N:
        right = mid - 1
    else:
        left = mid + 1

print(right)